File: stoei/slurm/nodelist.py
"""Slurm NodeList expression parser for expanding bracket notation to hostnames."""

from stoei.logger import get_logger

logger = get_logger(__name__)


def _split_nodelist(s: str) -> list[str]:
    """Split a NodeList string on commas that are not inside brackets.

    Args:
        s: A NodeList string like "node01,node[03-05],gpu[01-02]".

    Returns:
        List of individual NodeList tokens.
    """
    tokens: list[str] = []
    depth = 0
    current: list[str] = []

    for ch in s:
        if ch == "[":
            depth += 1
            current.append(ch)
        elif ch == "]":
            depth -= 1
            current.append(ch)
        elif ch == "," and depth == 0:
            tokens.append("".join(current))
            current = []
        else:
            current.append(ch)

    if current:
        tokens.append("".join(current))

    return tokens


def _expand_bracket_expr(expr: str) -> set[str]:
    """Expand a single bracket expression like "node[01-04]" or "node[01,03]".

    Args:
        expr: A bracket expression token, e.g. "node[01-04,07]".

    Returns:
        Set of expanded hostnames. Returns empty set if expression is malformed.
    """
    try:
        bracket_open = expr.index("[")
        bracket_close = expr.index("]")
    except ValueError:
        logger.warning(f"Malformed NodeList expression (missing bracket): {expr!r}")
        return set()

    prefix = expr[:bracket_open]
    spec = expr[bracket_open + 1 : bracket_close]

    result: set[str] = set()
    for part in spec.split(","):
        if "-" in part:
            range_parts = part.split("-", 1)
            try:
                start_str, end_str = range_parts[0], range_parts[1]
                start = int(start_str)
                end = int(end_str)
                # Preserve zero-padding width from the start token
                width = len(start_str)
                for i in range(start, end + 1):
                    result.add(f"{prefix}{str(i).zfill(width)}")
            except ValueError:
                logger.warning(f"Malformed range in NodeList expression: {part!r}")
        else:
            result.add(f"{prefix}{part}")

    return result


def expand_nodelist(nodelist: str) -> set[str]:
    """Expand a Slurm NodeList expression to a set of individual hostnames.

    Handles the full Slurm bracket notation including ranges, comma-separated
    lists within brackets, multiple prefix groups, and zero-padded indices.
    Returns an empty set for pending-state placeholders like "(None)" and
    for empty strings.

    Args:
        nodelist: A Slurm NodeList string, e.g. "node[01-04]",
            "node01,node[03-05]", "gpu[01-02],cpu[01-02]", or "(None)".

    Returns:
        Set of individual hostname strings. Empty set for pending placeholders,
        empty input, or malformed expressions (a warning is logged for the latter).

    Examples:
        >>> expand_nodelist("node01")
        {'node01'}
        >>> expand_nodelist("node[01-04]")
        {'node01', 'node02', 'node03', 'node04'}
        >>> expand_nodelist("(None)")
        set()
        >>> expand_nodelist("")
        set()
    """
    nodelist = nodelist.strip()

    if not nodelist or nodelist.startswith("("):
        return set()

    try:
        tokens = _split_nodelist(nodelist)
    except Exception as exc:
        logger.warning(f"Failed to parse NodeList {nodelist!r}: {exc}")
        return set()

    result: set[str] = set()
    for token in tokens:
        if not token:
            continue
        if "[" in token:
            result.update(_expand_bracket_expr(token))
        else:
            result.add(token)

    return result

File: internal/slurm/nodelist.go
package slurm

import (
	"fmt"
	"strconv"
	"strings"
)

// ExpandNodeList expands a Slurm NodeList expression into the set of individual
// hostnames it denotes, returned as a sorted, de-duplicated slice. It handles the
// bracket notation including ranges ("node[01-04]"), comma lists inside brackets
// ("node[01,03]"), multiple prefix groups ("gpu[01-02],cpu[01-02]"), and
// zero-padded indices. Pending-state placeholders such as "(None)" and empty
// input yield nil. Ports nodelist.expand_nodelist (which returns a Python set;
// here the result is sorted for determinism).
func ExpandNodeList(nodelist string) []string {
	set := expandNodeListSet(nodelist)
	if len(set) == 0 {
		return nil
	}
	out := make([]string, 0, len(set))
	for name := range set {
		out = append(out, name)
	}
	sortStrings(out)
	return out
}

// expandNodeListSet does the work of ExpandNodeList and returns the raw set so
// callers that only need membership or a count avoid an allocation/sort.
func expandNodeListSet(nodelist string) map[string]struct{} {
	nodelist = strings.TrimSpace(nodelist)
	if nodelist == "" || strings.HasPrefix(nodelist, "(") {
		return nil
	}

	result := make(map[string]struct{})
	for _, token := range splitNodeList(nodelist) {
		if token == "" {
			continue
		}
		if strings.Contains(token, "[") {
			for name := range expandBracketExpr(token) {
				result[name] = struct{}{}
			}
		} else {
			result[token] = struct{}{}
		}
	}
	return result
}

// splitNodeList splits a NodeList string on commas that are not inside brackets,
// porting nodelist._split_nodelist.
func splitNodeList(s string) []string {
	var tokens []string
	depth := 0
	var current strings.Builder
	for _, ch := range s {
		switch ch {
		case '[':
			depth++
			current.WriteRune(ch)
		case ']':
			depth--
			current.WriteRune(ch)
		case ',':
			if depth == 0 {
				tokens = append(tokens, current.String())
				current.Reset()
			} else {
				current.WriteRune(ch)
			}
		default:
			current.WriteRune(ch)
		}
	}
	if current.Len() > 0 {
		tokens = append(tokens, current.String())
	}
	return tokens
}

// expandBracketExpr expands one bracket expression such as "node[01-04,07]" into
// a set of hostnames, porting nodelist._expand_bracket_expr. A malformed
// expression (missing bracket or bad range) yields an empty set, matching the
// Python behavior of logging a warning and skipping.
func expandBracketExpr(expr string) map[string]struct{} {
	open := strings.IndexByte(expr, '[')
	closeIdx := strings.IndexByte(expr, ']')
	if open < 0 || closeIdx < 0 {
		return nil
	}

	prefix := expr[:open]
	spec := expr[open+1 : closeIdx]

	result := make(map[string]struct{})
	for _, part := range strings.Split(spec, ",") {
		if strings.Contains(part, "-") {
			rangeParts := strings.SplitN(part, "-", 2)
			startStr, endStr := rangeParts[0], rangeParts[1]
			start, err1 := strconv.Atoi(startStr)
			end, err2 := strconv.Atoi(endStr)
			if err1 != nil || err2 != nil {
				continue
			}
			width := len(startStr)
			for i := start; i <= end; i++ {
				result[fmt.Sprintf("%s%0*d", prefix, width, i)] = struct{}{}
			}
		} else {
			result[prefix+part] = struct{}{}
		}
	}
	return result
}
